### PR TITLE
IntelliJ project files

### PR DESCRIPTION
**Pull request by lrytz**
Fixes compilation within IDEA. Allows compiling and running JUnit tests directly withing the IDE.

Review by @retronym

### CODE REVIEW

File: src/intellij/setup.sh [hunk 1]
```diff
@@ -5,19 +5,9 @@
...
-	cp $f $NEW_FILE
-
-	# IntelliJ doesn't process the "compilerOptions" setting for variable
-	# replacement. If it did, we would just use "$PROJECT_DIR$". Instead,
-	# we do this replacement ourselves.
-	perl -pi -e 's/\$BASE_DIR\$/$ENV{"BASE"}/g' $NEW_FILE
-	echo "Created $NEW_FILE"
+for f in "$SCRIPT_DIR"/*.{iml,ipr}; do
+  cp $f.SAMPLE $f
 done
```
Comment (retronym) [inline src/intellij/setup.sh:14]: This part doesn't work from a clean checkout. I think it still needs to be:

``` bash
#!/usr/bin/env bash
#
# Generates IntelliJ IDEA project files based on the checked-in samples.
#

set -e
export SCRIPT_DIR="$( cd "$( dirname "$0" )" && pwd )"
echo "About to delete .ipr and .iml files and replace with the .SAMPLE files. Press enter to continue or CTRL-C to cancel."
read

(rm -f *.ipr *.iml 2>/dev/null)
for f in $(ls "$SCRIPT_DIR"/*.SAMPLE); do
    NEW_FILE=`echo $f | perl -pe 's/.SAMPLE//'`;
    cp $f $NEW_FILE
    echo "Created $NEW_FILE"
done
```

Comment (lrytz) [inline src/intellij/setup.sh:14]: I think it's fine, IntelliJ became smarter and identifies the path and inserts `$MODULE_DIR`, see https://github.com/scala/scala/pull/3708/files#diff-28f73def90f814ef684cfb478a660a53R8

Comment (retronym) [inline src/intellij/setup.sh:14]: But you can't loop over non-existent `"$SCRIPT_DIR"/*.{iml,ipr}` files, you have to loop over the `.SAMPLE` files and derive the `.iml` file name from that.

So we can remove `perl -pi -e 's/\$BASE_DIR\$/$ENV{"BASE"}/g' $NEW_FILE`, but still need the rest.

Comment (lrytz) [inline src/intellij/setup.sh:14]: Aw right, I copied that from the `update.sh` file, I intended to loop over the `.SAMPLE` files.

**Comment by retronym** [comment]
Other than that, LGTM!

I tested it out with an external Scala compiler server using IDEA's incremental compilation. This isn't as accurate as SBT's algorithm (also available in IntelliJ), but might work well for compiler development. Adding a new method to a supertype of Global, for instance, still gives about a 6s turnaround time.

Preferences, IDE Settings, Scala

image

**Comment by retronym** [comment]
I wonder if we could try unifying the output directories for Ant and IntelliJ. We could do this by symlinking, e.g,

`out/production/compiler -> build/quick/classes/compiler`

Then we could use the same class files for the command line `partest` or `build/qbin/bin/*`

We can experiment with this as a followup.

**Pull request merged**